docs/dev/framework/filesystem/virtual-filesystem.md

```diff
@@ -0,0 +1,292 @@
+---
+title: "Virtual filesystem"
+description: Extra metadata from the database for your file resources.
+aliases:
+  - /framework/filesystem/dbafs/
+---
+
+{{< version "4.13" >}}
+
+{{% notice warning %}}
+The new filesystem capabilities are currently considered *experimental* and therefore not covered by Contao's BC
+promise. Classes marked with `@experimental` should be considered internal for now. Although not likely, there could
+also be some behavioral changes, so be prepared.
+{{% /notice %}}
+
+
+## Names and autowiring
+Each virtual filesystem instance has a unique name which is used to generate a [named autowiring alias][NamedAliases].
+If you want to work with the `documents` filesystem, you would inject a variable named `$documentsStorage` (note the
+suffix 'Storage'):
+
+```php
+class Example {
+    public function __construct(private VirtualFilesystemInterface $documentsStorage) {}
+}
+```
+
+{{% notice tip %}}
+If you cannot use autowiring, you would explicitly inject the `contao.filesystem.virtual.documents` service instead.
+{{% /notice %}}
+
+
+## Working with files and directories
+
+### Example
+Let's build a simple content element, that displays the contents of a single directory in the filesystem. Our goal is to
+list *files only* and show their *path*, *file size* and - if set in the back end file manager - metadata *title*.
+
+For this example, we're using the virtual filesystem named `files` that is already configured by default in Contao. If
+you want to play along, this will be a good starting point: 
+
+```php
+<?php
+// src/App/Controller/FilesListController.php
+namespace App\Controller;
+
+use Contao\ContentModel;
+use Contao\CoreBundle\Controller\ContentElement\AbstractContentElementController;
+use Contao\CoreBundle\DependencyInjection\Attribute\AsContentElement;
+use Contao\CoreBundle\File\Metadata;
+use Contao\CoreBundle\Filesystem\VirtualFilesystemInterface;
+use Contao\Template;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+#[AsContentElement(category: 'files')]
+class FilesListController extends AbstractContentElementController
+{
+    // We inject the 'files' virtual filesystem instance, that is scoped to the
+    // 'files' directory.
+    public function __construct(private VirtualFilesystemInterface $filesStorage) {}
```

Comment:
Same here. We should probably discuss the minimum PHP version that the examples in the documentation should use. Always cutting edge or lowest supported.

Reply:
I think we could start promoting PHP8 in the dev docs. In case of the filesystem you should be encouraged to use it anyway, so that you get the real `directoryExists()` :stuck_out_tongue:.

But ultimately up to you. I'd also be fine with rewriting it or adding a second tab.

Reply:
Yeah, I am not sure yet. On the one hand I want readers to be able to just copy and paste the code. But since Contao 4.9 and 4.13 still support PHP 7, it would not be convenient for those using PHP 7. I mean ideally we would have both versions, but that's a lot more work of course 😁 

I suppose we could distinguish between code that should be copy-able as is (like hooks and callbacks) and code that are just examples.

Reply:
see 17129a2ebfc934ef3c35d5c0538d9455dc4ada6f